design/back.compatibility.md

```diff
@@ -0,0 +1,79 @@
+
+## Design Options
+
+I considered two basic approaches to implementing multi-generation support
+1. conditional checks sprinkled in whenever changes are applied("if (generation==2) then do A else do B". Pro: it's easy to understand the concept, and easy to start. Con: extremely error prone and hard to test.
+2. complete copy of the generation-specific code (i.e. almost all of the code in the dll). Pro: Minimizes risk of errors. Con: May bloat the plugin size, source tree size, and compile time. Still have to watch out for changes in the wrong version.
+
+The obvious choice here is 2. Approach 1 would causes things to break often for no reason, no matter how careful developers are.
+
+### Source Code Management
+Here are 2 options to managing the source code
+1. move all the generation-specific code into a subfolder per generation - this is where the preview generation is worked on. When finalizing a generation and starting a new preview generation, create a copy of the preview generation folder. Change the namespace in the subfolder. Add a generation-specific prefix on each source file's filename
+2. move all generation-specific code into a separate repository, and have the main repository include this repository as a submodule for each generation.
+
+* Pros for 1: Easy to use with git. 
+* Cons for 1: It's clunky to have a prefix on filenames (but makes it much harder to accidentally modify things in a final generation - many tools highlight filename but not path). It's still possible to modify the wrong generation's files in the repo.
+* Pros for 2: Makes it harder to do accidental checkins in the wrong repo - you commit into a specific version.
+* Cons for 2: git submodule workflow is clunky, not integrated well with visual git tools. While it seems like meging across generations is easier, it's not something we'll see in practice as fixing finalized generations is not something we should be doing.
+
+I recommend version 1. The pain of the git submodule workflow, and loss of productivity is not worth the minor benefits of 2
+
+### Final Binaries
+Here are two options for building the the binaries
+1. Build a single DLL that includes all generations, with a small bit of generation-shared code for connecting to a server, finding the server version, and choosing the generation-specific mre. 
+2. Build one DLLs per generation, plus a management DLL. 
+Pros for 1: it's easier to have a single output DLL than multiple, as build scripts don't need to be updated when a generation is added. Total binary size is smaller than 2, as compiler will probably optimize out many functions that are unchanged across generations. Con: Higher build time for the DLL.
+I recommend option 1. 
+
+
+### Summary of Recommended Design
+We'll duplicate the files into a different subfolder per generation, add a generation-specific prefix to each file, and leave only the absolutely minimal set of files outside this folder. All of the code will be built into a single DLL.
```

Comment:
>built into a single DL [](start = 212, length = 22)

I really hope we can just use "major versions" instead of "generations". Having a .0 version be "finalized" seems very strange from a semver perspective. My proposed alternative: patches only accumulate in the latest major version. So 1.3.17 could be finalized when we cut 2.0.0.

Reply:
Yeah, I'll be updating design based on this feedback, so that a final version for generation 1 is 1.n instead of 2.0

Reply:
The terms "major version" and "generation" are slightly different to me. Major Version 3 would include generation 1, generation 2, and generation 3 code. I use this term, because "major version" is a generally understood, and I want to make sure any new dev coming in would need to understand the difference.